plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/lsp/AmazonQLspService.kt

```diff
@@ -195,6 +235,34 @@ class AmazonQLspService(private val project: Project, private val cs: CoroutineS
         instance = start()
     }
 
+    private suspend fun waitForRestartSlot(): Unit = restartMutex.withLock {
+        val currentTime = System.currentTimeMillis()
+
+        while (restartTimestamps.isNotEmpty() &&
+            currentTime - restartTimestamps.first() > RESTART_WINDOW_MS
+        ) {
+            restartTimestamps.removeFirst()
+        }
+
+        if (restartTimestamps.size < MAX_RESTARTS) {
+            restartTimestamps.addLast(currentTime)
+            return
+        }
+
+        val oldestTimestamp = restartTimestamps.first()
+        val waitTimeMs = (oldestTimestamp + RESTART_WINDOW_MS) - currentTime
+
+        LOG.info { "Rate limit reached for LSP server restarts. Waiting ${waitTimeMs}ms for next available slot." }
```

Comment:
if the limit is exceeded, should we just assume that there is an issue with the process and give up?

Reply:
I can change it to give up trying